Full V13 support: Interactions (buttons and select menus) and reactions!

psibean opened the pull request:
I have effectively abandoned this PR and released my own, full v13 compatible version.

[Find it here](https://github.com/psibean/discord.js-pagination) with it's documentation and a working example bot.

Install with

```
npm install @psibean/discord.js-pagination
```

psibean committed:
Replace unary operators

psibean committed:
Add custom footer and page resolvers, shift all optional and default arguments into an options object.

psibean committed:
Ignore vscode

psibean committed:
Remove the now redundant emojiList length check

psibean committed:
Expose user on the collect event, await the reaction removal.

psibean committed:
Remove unnecessary footerResolver param comment line

psibean committed:
Prevent crash if message is deleted before reactions are applied.

psibean committed:
Always filter against bots, always ensure emoji is in list, apply custom filter only if it exists.

psibean committed:
Shift collection reactor setup prior to the bot adding reactions.

psibean committed:
Expose the message to the pageResolver

psibean committed:
Mark resolvers optional in typings

psibean committed:
Pass in the correct message to the pageResolver

psibean committed:
Fix emoji list typing

psibean committed:
Additional page message delete guard

psibean commented:
If you do something like this with this code:

```
    const pageResolver = async (message, pages, emojiList, currentPageIndex, reaction) => {
      let newPage = currentPageIndex;
      switch (reaction.emoji.name) {
        case emojiList[0]:
          newPage = currentPageIndex > 0 ? currentPageIndex - 1 : pages.length - 1;
          break;
        case emojiList[1]:
          newPage = currentPageIndex + 1 < pages.length ? currentPageIndex + 1 : 0;
          break;
        case emojiList[2]:
          await message.delete();
          break;
        default:
          return newPage;
      }
      return newPage;
    };
    const emojiList = ['⏪', '⏩', '🇽']

    return paginationEmbed(message, pages, { emojiList, pageResolver });
```

If you click the x to delete the message and also click an arrow before the message is deleted, there's a race condition on `await reaction.users.remove(user.id);` in the collect event  with the removal of the users react and the deletion of the message. A better way to handle this would be to queue the collected reactions and then handle them from the queue in the order they are received. I'll see if I can do this as well at a later time.

Edit: This is a discord.js race condition issue.

Alcremie commented on index.d.ts:
```diff
@@ -0,0 +1,14 @@
+  deleteOnEnd?: boolean,
+  useUtil?: boolean,
+  emojiList?: EmojiIdentifierResolvable[],
+  collectorFilter?(reaction: MessageReaction, user: User): boolean | Promise<boolean>
```
This should at least be consistent and also have a trailing `,`.

Alcremie commented on index.d.ts:
```diff
@@ -0,0 +1,14 @@
+  emojiList?: EmojiIdentifierResolvable[],
+  collectorFilter?(reaction: MessageReaction, user: User): boolean | Promise<boolean>
+  footerResolver?(pageIndex: number, pagesLength: number): string,
+  pageResolver?(msg: Message, pages: MessageEmbed[], emojiList: string[], currentPageIndex: number, reaction: MessageReaction): number | Promise<number>
```
Ditto here with the trailing comma.

Alcremie commented on index.d.ts:
```diff
@@ -0,0 +1,14 @@
+  pageResolver?(msg: Message, pages: MessageEmbed[], emojiList: string[], currentPageIndex: number, reaction: MessageReaction): number | Promise<number>
+}
+
+declare function paginationEmbed (msg: Message, pages: MessageEmbed[], paginationOptions?: PaginationOptions): Promise<Message>
```
Trailing semicolon here for consistency, as well as removing the space between `paginationEmbed` and its parameters.

Alcremie commented on index.js:
```diff
@@ -1,33 +1,81 @@
+ * @param {PaginationOptions} paginationOptions - exposes collector options, provides customization.
+ */
+const paginationEmbed = async (
+										msg, pages,
```
I'm not sure what happened here with the indentation...

Alcremie commented on index.js:
```diff
@@ -1,33 +1,81 @@
-		(reaction, user) => emojiList.includes(reaction.emoji.name) && !user.bot,
-		{ time: timeout }
+		(reaction, user) => emojiList.includes(reaction.emoji.name) && !user.bot
+							&& (collectorFilter ? collectorFilter(reaction, user, emojiList) : true),
```
Ditto here with the borked indentation.

Alcremie replied:
Also, `collectorFilter` should both default to `() => true` and not `null`, and it should be awaited, since you type it as returning either `boolean` or `Promise<boolean>`.

Alcremie commented on index.js:
```diff
@@ -1,33 +1,81 @@
+	return newPage;
+};
+
+const defaultFooterResolver = (currentPageIndex, pagesLength) =>`Page ${currentPageIndex + 1} / ${pagesLength}`;
```
Space after `=>` for consistency.

Alcremie commented on index.js:
```diff
@@ -1,33 +1,81 @@
 	const reactionCollector = curPage.createReactionCollector(
-		(reaction, user) => emojiList.includes(reaction.emoji.name) && !user.bot,
-		{ time: timeout }
+		(reaction, user) => emojiList.includes(reaction.emoji.name) && !user.bot
```
It should probably be left to the user whether or not they want to listen to bot reactions for whatever reason.

Alcremie commented on index.js:
```diff
@@ -1,33 +1,81 @@
+	try {
+		for (const emoji of emojiList) await curPage.react(emoji);
+	} catch (error) {
+			reactionCollector.stop();
```
The reaction collector will automatically stop if the message it references was deleted, so there's no need for an explicit `try`/`catch` here; just attempt to add the reactions and let any `10008` unknown message errors propagate as usual.

Alcremie commented on index.js:
```diff
@@ -1,33 +1,81 @@
-		curPage.edit(pages[page].setFooter(`Page ${page + 1} / ${pages.length}`));
+	reactionCollector.on('collect', async (reaction, user) => {
+		// In case a react deletes the curPage msg in the pageResolver and another react is quickly made.
+		if (curPage.deleted) return;
```
Ditto here, this doesn't need to be checked.

Alcremie reviewed (commented):
Just an initial sweep of the PR, since I got the notification from the issue that one was created.

Alcremie commented on index.js:
```diff
@@ -1,33 +1,81 @@
-	const curPage = await msg.channel.send(pages[page].setFooter(`Page ${page + 1} / ${pages.length}`));
-	for (const emoji of emojiList) await curPage.react(emoji);
+	pages[page].setFooter(footerResolver(page, pages.length));
+	const curPage = useUtil ? await msg.util.reply(pages[page]) : await msg.channel.send(pages[page]);
```
I'm not exactly sure how I feel about this - maybe you should just specify this as a function that accepts a `MessageEmbed` and returns a `Promise<Message>` so that you're not adding specific support for only one framework over another.

psibean commented:
Thanks for the feedback, completely agree with all points. I’ll get back into it and resolve / reply more specifically once I’m done with work!

psibean committed:
Typing syntax ',' to ';', add sendMessage property of PaginationOptions.

psibean committed:
Remove uinnecessary deleted check and try/catch, add defaultCollectorFilter and sendMessage option.

psibean commented:
- Switched all `,` to `;` in typings and made consistent.
 - Fixed spacings and attempted to fix indentation (it wasn't appearing like that on my side).
 - Set `defaultCollectorFilter` to retain the original logic for increased backwards compatibility.
 -  Added a sendMessage option - I always confuse my overlap of what is discord.js and what is akairo.
 - Removed the redundant try / catch and stop and extra deleted check.
 - Switched the `deleteOnEnd` check on `end` to use `Message#deletable`

In my previously presented example,
```
await reaction.users.remove(user.id);
```
this line still throws an unknown message error if you hit a delete option then react another before the message is deleted. Since this is happening in the collect event, it's not something you can catch.

I discussed the `Message#deleted` issue on discord.js discord server [here](https://discord.com/channels/222078108977594368/682166281826598932/801026030349189120) - it's a troublesome race condition, for my use case I'll work around it.

But it does mean there is still potential for errors to be thrown from the reaction removal in a case where collected reactions are sent shortly after a manual message delete is made, but before the message_deleted gateway response has come back for the message to be marked as deleted and fire the messages deleted event.

Alternatively, I could pass the user to the pageResolver and handle the removal there, then people can also provide their custom logic and error handling around that.

psibean committed:
Add missing async, fix paghe > page typo

Alcremie commented on index.js:
```diff
@@ -21,60 +21,54 @@ const defaultPageResolver = async (_, pages, emojiList, currentPageIndex, reacti
-const defaultFooterResolver = (currentPageIndex, pagesLength) =>`Page ${currentPageIndex + 1} / ${pagesLength}`;
+const defaultFooterResolver = (currentPageIndex, pagesLength) => `Page ${currentPageIndex + 1} / ${pagesLength}`;
+
+const defaultSendMessage = async (message, pageEmbed) => await message.channel.send(pageEmbed);
```
Due to the way Promises work, you don't even have to mark this as `async` and use an explicit `await`!
```ts
async function defaultSendMessage(message: Message, pageEmbed: MessageEmbed): Promise<Message> {
    return await message.channel.send(pageEmbed);
}
```
And
```ts
function defaultSendMessage(message: Message, pageEmbed: MessageEmbed): Promise<Message> {
    return message.channel.send(pageEmbed);
}
```
End up being pretty much exactly the same, as shown via the return type. The only thing you're doing in the first one is unboxing the `Promise<Message>` into the inner `Message` and then reboxing it into another `Promise` due to the `async` function, and in the second, you're just directly returning that `Promise<Message>`. They should both probably return the same error stack should something reject, but I could double check if you'd like; I know that discord.js made its error messages significantly more readable in v12 when you explicitly awaited.

psibean replied:
Ah right, duh, it's being awaited already at the top level! I just keep coming back to this after ~9 hours of work.

Alcremie commented on index.js:
```diff
@@ -21,60 +21,54 @@ const defaultPageResolver = async (_, pages, emojiList, currentPageIndex, reacti
-										} = {}) => {
+const paginationEmbed = async (msg, pages,
+	{
+		deleteOnEnd = false, useUtil = false,
```
I would probably prefer these being on separate lines, just for consistency. :)

Alcremie commented:
I think the best thing regarding reaction removal would be to just add a `Promise#catch` noop, e.g.
```js
await reaction.users.remove(user.id).catch(() => {});
```
Or otherwise allow it to propagate and let the user handle it themselves. I've thought about proposing a throttle for message editing, so that deleting a message via a reaction and then immediately trying to react should not cause an edit to happen from the time the message is deleted to the time the reaction collector catches up and automatically ends, but this doesn't solve the problem of trying to react immediately after the message was deleted through other means (e.g. manually by another user).

psibean committed:
Rename page > currentPageIndex

psibean committed:
Rename curPage > paginatedEmbedMessage

psibean committed:
Remove useUtils from PaginationOptions, type parameter name clarifications.

psibean committed:
Add collectErrorHandler

psibean committed:
Update example usage

psibean committed:
Remove redundant useUtil

psibean committed:
Remove unnecessary explicit async/await.

psibean commented:
> I think the best thing regarding reaction removal would be to just add a `Promise#catch` noop, e.g.
> 
> ```js
> await reaction.users.remove(user.id).catch(() => {});
> ```
> 
> Or otherwise allow it to propagate and let the user handle it themselves. I've thought about proposing a throttle for message editing, so that deleting a message via a reaction and then immediately trying to react should not cause an edit to happen from the time the message is deleted to the time the reaction collector catches up and automatically ends, but this doesn't solve the problem of trying to react immediately after the message was deleted through other means (e.g. manually by another user).

Putting a `.catch()` on the `paginationEmbed` call doesn't catch any of the errors on the `collect` listener, as the `paginationEmbed` call is resolved with the return of the sent message. I'm not familiar with how to propagate those errors in a way they can otherwise be caught?

Not sure if it's appropriate, but for now I've wrapped all the `collect` event handlers logic in a try / catch and passed the error through to a `collectErrorHandler`.

I also updated the readme example!

Should I pass through the `reactionCollector` and `paginatedEmbedMessage` to the error handler so users can make a decision on what to do with them based on the error?

Edit: Ugh, I don't know why it's giving weird indentation issues, I checked my vscode settings, it shouldn't be doing that.

psibean committed:
Re-add MessageEmbed require to example

psibean committed:
Fix indentation

psibean committed:
Fix indentation

psibean committed:
amend merge

psibean committed:
Update index.js

psibean committed:
Fix example usage syntax error

psibean committed:
Merge branch 'master' of https://github.com/psibean/discord.js-pagination into master

psibean committed:
Add deleted guard on end (in case ended by other means)

psibean committed:
Destructure params - optional, delegate end handler, expose messages to error handler.

psibean committed:
Update error handler example

psibean committed:
Fix collector filter typing, adjust consistent syntax

ikrishagarwal reviewed (commented):
Awesome Upgrade. Loved it.

DudeThatsErin commented:
This is not working for me.

First, if I include `EmojiList` in the options on the help.js, page it states "emojiList is not defined" do I put emojies inside brackets there?

Second, this is all of the code: https://sourceb.in/W6WKao4Cvk
I'm getting cannot send an empty message.

psibean commented:
Apologies if formatting here isn't quite right, I'm on mobile.

> First, if I include `EmojiList` in the options on the help.js, page it states "emojiList is not defined" do I put emojies inside brackets there?

This error means that the actual variable emojiList, which you're passing into the function does not exist where you're using it. This is to do with variable scope.

Example:

```
if (someCondition) {
  const emjiList = [];
// can only access emojiList in this block
  ...
} else {
  // emojiList is not defined here
}
// emojiList is not defined here

```

Make sure whatever you pass to the paginationEmbed function is actually within scope of the call.

> Second, this is all of the code: https://sourceb.in/W6WKao4Cvk
> 
> I'm getting cannot send an empty message.

This ones a bit trickier. My initial glance at the code, nothing jumps out. From what I can see, you should have at least 10 valid pages. Does this happen right away, or after navigating? The error indicates that your array has a blank / empty embed, or a null/undefined value

Happy to help over discord!

Also please be aware this version does not yet work with djs v13 which was just released.

If you're using djs b13 this WILL be a problem.

I will update it to v13 in a few hours!

I may also publish this version to npm to make it easier for people to use

DudeThatsErin commented:
> Apologies if formatting here isn't quite right, I'm on mobile.
> 
> > First, if I include `EmojiList` in the options on the help.js, page it states "emojiList is not defined" do I put emojies inside brackets there?
> 
> This error means that the actual variable emojiList, which you're passing into the function does not exist where you're using it. This is to do with variable scope.
> 
> Example:
> 
> ```
> if (someCondition) {
>   const emjiList = [];
> // can only access emojiList in this block
>   ...
> } else {
>   // emojiList is not defined here
> }
> // emojiList is not defined here
> ```
> 
> Make sure whatever you pass to the paginationEmbed function is actually within scope of the call.
> 
> > Second, this is all of the code: https://sourceb.in/W6WKao4Cvk
> > I'm getting cannot send an empty message.
> 
> This ones a bit trickier. My initial glance at the code, nothing jumps out. From what I can see, you should have at least 10 valid pages. Does this happen right away, or after navigating? The error indicates that your array has a blank / empty embed, or a null/undefined value
> 
> Happy to help over discord!
> 
> Also please be aware this version does not yet work with djs v13 which was just released.
> 
> If you're using djs b13 this WILL be a problem.
> 
> I will update it to v13 in a few hours!
> 
> I may also publish this version to npm to make it easier for people to use

Ah, yeah, I'm using DJS V13. My discord is DudeThatsErin#8061 whenever you get this working with v13, please shoot me a message. I would definitely appreciate it. Appreciate the fast response as well.

psibean commented:
> Ah, yeah, I'm using DJS V13. My discord is DudeThatsErin#8061 whenever you get this working with v13, please shoot me a message. I would definitely appreciate it. Appreciate the fast response as well.

I have since updated this as per the [discordjs Updating from v12 to v13 guide](https://discordjs.guide/additional-info/changes-in-v13.html). I have also added you on discord, please let me know if there are any issues with the updates, I'll be happy to continue support and maintenance!

I have also published my version of the package to npm. You can now install and use via:
```
npm install @psibean/discord.js-pagination
```
```
const paginationEmbed = require('@psibean/discord.js-pagination');
```
Edit: Had not yet pushed the v13 changes into the branch for this PR - they're in now.

For those interested, the changes in v13 that impact this package are:
```
message.channel.send(embed)
message.edit(embed)
```
become
```
message.channel.send({ embeds: [embed] });
message.edit({ embeds: [embed] });
```
respectively, and for the message#createReactionCollector call, the reaction collector filter callback gets passed in as filter in the options.


On this PR I bumped the version to 2.0.0 as it's no longer backwards compatible.

If you want this version of the code which is still compatible with discordjs v12, use `@psibean/discord.js-pagination@1.1.0`

psibean committed:
Update with initial discordjs v13 support

psibean committed:
Update index typings

psibean committed:
Update readme installation and version support information

psibean committed:
Bump major

psibean committed:
Fix async/await defaults, fix collectorFilter destruct params

psibean committed:
Add missing filter return, add interaction typing

psibean commented:
There were some additional issues found, they have now been fixed and this has been tested with discord.js v13 working as of v2.0.8. The paginationEmbed can take either a message or an interaction. (only used to access `Message/Interaction#channel#send`, unless you otherwise customise sendMessage).

The fixes were also ported to the discord.js v12 version of the package which is now working at v1.1.1

psibean commented:
**I'm in the process of updating this to be class based and to support MessageComponents (buttons and select menus).**
I currently have support for buttons, select menus and message reactions.
The buttons and select option paginations provide a "custom prefix" so you can avoid the interactions in your client ineractionCreate, e.g.: 
```
if (interaction.customId.startsWith('pagination')) return
```
Each pagination also uses the received interaction/messages id (from the command initiating the pagination) as a suffix. This way in each paginations collectorFilter you can do:
```
interaction.customId.endsWith(paginator.customIdSuffix)
```

This allows all of the event listeners to easily react to only the interactions they need.

WIP: https://github.com/psibean/discord.js-pagination/pull/1

The PR branch contains a WORKING example too, I also plan on fully documenting once it's ready.

psibean commented:
I've abandoned the PR at this point.

Find the new version [over here](https://github.com/psibean/discord.js-pagination).

Complete support for reaction, button and select menu pagination. Fully customizable and controllable.
